Layout and placement

stefanhannie opened the pull request:
Here is the initial spring layout implementation. At a high level it's dead simple:
- Compute the layouts for `S` and `T`
- Map vertices of `S` to the "closest" vertices of `T`
- Turn that mapping into `initial_chains`
- Hand those `initial_chains` to `minorminer`

Commit:
Initial commit for layout and placement

Includes kamada kawai layout and various initial chain placement options

Commit:
Bug Fix

A stray " and some missing args.

Commit:
Documentation

Added more documentation and formatting

Commit:
Change structure of contribution

This serves for easier integration into minorminer.

Commit:
Bug Fixes

Import errors and typos. Added a helper function for find_embedding.

Commit:
Fix a silly injective requirement

Vertices in T are renamed when making the assignment problem

Commit:
Neighborhood extension bug

Was changing the set as I was iterating through it.

Commit:
Structural changes

The parsing of kwargs and folded extend in as a construction option.

Commit:
Save computation

You can now pass in precomputed layouts.

boothby committed:
added layout submodule to setup.py

Commit:
Edges as input graphs

Can accept edges, as well as NetworkX graphs, as input.

Commit:
Injective algorithm refactor

Cleaner and more robust against vertex naming conflicts.

Commit:
Unit tests for layout

Tests the Kamada Kawai layout for several dimensions.

Commit:
Added Chimera layout

The Chimera layout --> [-1,1]^2 was added to standardize layouts.

Commit:
High dimension Chimera layout

A nontrivial 3d Chimera layout implementation

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+
+def find_embedding(S, T, layout=kamada_kawai, placement=closest, construction=singleton, hinting=initial, **kwargs):
+    """
+    Tries to embed S in T by computing layout aware initial_chains and passing them to minorminer.
```
initial_chains --> this can be made into a reference to the parameter, then users can click to it
"computing layout aware" should be "computing layout-aware"
"passing them to minorminer" --> we're in minorminer. Do you mean the previous find_embedding? Does this function need a different name to distinguish it from that one?

stefanhannie replied:
* I'm okay renaming the function. @boothby what do you think about `layout_find_embedding()`? A bit long, but it should just be a temporary name.
* How do I make a reference to something like `initial_chains`?

boothby replied:
I like the function having the same name.  It appears in a submodule for the purpose of differentiation.  As for additional parameters, I'd simply refer the reader to the documentation for minorminer.find_embedding.

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        The graph you are embedding (source) or a NetworkX supported data structure (see to_networkx_graph()).
+    T : NetworkX graph or edges data structure (dict, list, ...)
+        The graph you are embedding into (target) or a NetworkX supported data structure (see to_networkx_graph()).
+    layout : function or (function/dict, function/dict) (default kamada_kawai)
```
Will be cvlearer if you move the default to the end of the description instead. "S and the second applies to T. Default is kamada_kawai."

stefanhannie replied:
Are you not worried about breaking the convention that all other keyword arguments are 
```arg_name : arg_type (default default_arg)```?

JoelPasvolsky replied:
I afterwards noticed that minorminer uses this different convention from other Ocean tools.

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+    T : NetworkX graph or edges data structure (dict, list, ...)
+        The graph you are embedding into (target) or a NetworkX supported data structure (see to_networkx_graph()).
+    layout : function or (function/dict, function/dict) (default kamada_kawai)
+        Will either compute a layout or pass along precomputed layouts. If it's a single object, it has to be a function
```
"Will either compute a layout or pass along precomputed layouts." --> I think what you're saying here is that this parameter ""Specifies either a single function to compute the layout for both S and T or two pre-computed layouts, the first for S and the second for T."

stefanhannie replied:
Almost, it's a bit more general. It can be a single function, or a tuple with any combination of functions and layouts. i.e. `(f(), {})` is fine. What about:

"Specifies either a single function to compute the layout for both S and T or a 2-tuple consisting of functions or pre-computed layouts, the first applying to S and the second applying to T."

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        Will either compute a layout or pass along precomputed layouts. If it's a single object, it has to be a function
+        and it applies to both S and T. If it's a tuple of objects the first applies to S and the second applies to T.
+    placement : function (default closest)
+        The placement algorithm to call.
```
Could you add a bit more here?

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+    Returns
+    -------
+    emb : dict
+        Output is dependant upon kwargs passed to minonminer, but more or less emb is a mapping from vertices of 
```
dependent

JoelPasvolsky commented on minorminer/layout/__init__.py:
```diff
@@ -0,0 +1,118 @@
+
+def parse_kwargs(kwargs):
+    """
+    Pull out kwargs for layout and construction functions. Leave the remaining ones for minorminer.
```
Consider replacing "Pull out" with "Extract", we like to be fancy here

JoelPasvolsky commented on minorminer/layout/construction.py:
```diff
@@ -0,0 +1,122 @@
+
+def singleton(S, T, placement, extend=False):
+    """
+    Given a placement (a map phi from vertices of S to vertices of T), form the chain [phi(u)] for each u in S.
```
a map, ``phi``, from vertices

stefanhannie replied:
Is the comment here to add: 
* commas (I should have done that)
* write `phi` (with back-ticks)
* :math: `phi` (use the `:math:` prefix thing I've seen in other places in the project)

JoelPasvolsky commented on minorminer/layout/construction.py:
```diff
@@ -0,0 +1,122 @@
+
+def neighborhood(S, T, placement, second=False, extend=False):
+    """
+    Given a placement (a map phi from vertices of S to vertices of T), form the chain N_T(phi(u)) (closed neighborhood 
```
a map, ``phi``, from vertices

JoelPasvolsky commented on minorminer/layout/placement.py:
```diff
@@ -0,0 +1,60 @@
+
+def injective(S_layout, T_layout):
+    """
+    Injectively maps vertices of S to the closest vertices of T as given by S_layout and T_layout. i.e. This is the 
```
"and T_layout. i.e. This is" --> either "and T_layout; i.e. this is" or else "and T_layout. This is"

JoelPasvolsky reviewed (approved):
LGTM. One thing that confuses me is why is the find_embedding() in file layout/__init__.py? That seems unusual to me. Is it temporary and then going through a renaming?

boothby commented:
> LGTM. One thing that confuses me is why is the find_embedding() in file layout/**init**.py? That seems unusual to me. Is it temporary and then going through a renaming?

This was my recommendation, actually -- it's not _terribly_ unusual but the function may move at a later date.

Commit:
Documentation changes

Feedback dictated documentation changes. Minor typo fixes.

Commit:
Placement unit tests

Added unit tests for placement algorithms. Cleaned up test struture.

Commit:
Streamlined layout center and scale

Parameters center and scale are handled in a more systematic way.

Commit:
Scale layouts by edge length

It may be desirable to have layouts with matching edge lengths

Commit:
Refactor layout

Created Layout class. This consolidates boilerplate code.

stefanhannie committed:
Added install step

Cannot find module _minorminer without the build_ext --inplace step.

stefanhannie committed:
Binning as placement

Working version of a binning placement algorithm

Commit:
Layout bug fix

Wasn't passing in seed correctly; data structure changed for unit tests.

Commit:
Scale bug fix

Check that scale vectors passed in are of the correct dimension.

Commit:
Chimera specific binning

An algorithm that uses the coordinates of a chimera graph for binning

Commit:
Refactor binning

Removed is_chimera flag. Moved dnx specific things to utils.

Commit:
Refactor integer lattice

Added some helper functions and a unit test.

Commit:
Better closer placement

Faster (uses a KDTree) and now allows maps to connected subgraphs.

JoelPasvolsky committed:
Update conf.py and local references

JoelPasvolsky committed:
Update intersphinx refs

JoelPasvolsky committed:
Prevent build warning

JoelPasvolsky committed:
Prevent duplicates warnings on sdk_index ref

Commit:
New default binning strategy

Extend a unit larger than desired for better converage when rounding.

Commit:
PCA layout implemented

A quick n dirty implementation of the PCA layout method.

JoelPasvolsky committed:
Fix dwavebinarycsp --> minorminer

Commit:
Scale and center PCA

The PCA algorithm needed to match the extent of the other algorithms.

Commit:
PCA bug fix

The next pivot should be far from all previous pivots.

Commit:
Rescalling changes

Fixed rescalling bug. Added seed and unit test for PCA.

JoelPasvolsky committed:
Fix warnings on include (except matlab)

stefanhannie committed:
Gridify pegasus

Using the nice coordinates of pegasus, map vertices to lattice points.

JoelPasvolsky committed:
Add version notice

Commit:
Rotate layouts

Find minimum bounding box and rotating to align layouts

Commit:
Rotation unit tests

Fixed unit testing for rotating layouts.

Commit:
Encourage nonoverlapping chains

Modified the closest placement strategy to penalize overlap.

Commit:
Custom metric embedding (WIP)

A Kamada-Kawai-eque algorithm for custom metric spaces and functions

Commit:
Fast graph distance approximation

A PCA inspired algorithm for computing approximate graph distances.

Commit:
Cross placement

Each chain span the interval of the neighborhood from the layout.

Commit:
"T" placement strategy

Similar to the cross strategy, but more judicious in its chain growth.

Commit:
l1 gradient implemented

It's still a bit slow, but it works.

Commit:
Slightly faster l1 gradient

A slightly more vecetorized version

Commit:
Fully vectorized gradient

3d matrices are weird.

Commit:
Einsum l1 gradient

An even faster gradient.

Commit:
Plane x Torus gradient

A custom metric space for Chimera (and maybe Pegasus).

Commit:
p-norm layout

An efficient algorithm for a kamada-kawai-esque p-norm algorithm.

Commit:
Added row/column placement

This is pulled from crosses and tees which will be moved to construction

Commit:
Crosses as construction (WIP)

Abstracting the X-strategy as a construction for arbitrary layouts

Commit:
Bug fixes

Various small fixes

Commit:
Intersection knows about size

This is the first size aware scheme--this allows for direct emb to C(16)

Commit:
Refactoring

Removed Kamada-Kawai and made rescaling/recentering/etc. more cohesive

Commit:
Added requirements and bug fixes

Fixes in rescaling p_norm and unittests.

arcondello commented:
Looks like cmake build issue is probably related to https://github.com/google/googletest/issues/2711

Commit:
Injective intersection placement

This uses a topple approach. A more robust algorithm is desirable,
i.e. seam carving

Commit:
Random removal hinting strategy

Randomly removes some percentage of qubits from chains before passing to minorminer

Commit:
A cleaner injective_intersection

Using an actual nx.grid_graph as the data structure.
Topple algorithm is better as well

Commit:
Added some random noise to topple

Picks a random hungriest neighbor to topple to.

Commit:
Keep ordering but randomly map vars in a unit cell

This is a stopgap until the more robust placement is worked out.

boothby committed:
remove python 2.7 and 3.4 support

boothby committed:
remove python 2.7 and 3.4 build/deploy

stefanhannie committed:
Python3.5 compatibility

Changed f-strings to .format(); minor bug fix for pca.

boothby committed:
add mm.layout.utils package to setup.py, localize imports

stefanhannie committed:
Minor changes

Renamed a parameter and pulled a missing kwarg.

stefanhannie committed:
Refactored injective_intersection into binning

Consolidated some placement algorithms as options of each other.
Removed legacy code in favor of cleaner newer code.

stefanhannie committed:
Remove unused code; better unittests

Removed R2xT layout and supporting functions.
Removed other unused functions.
Unit tests are more robust.

stefanhannie committed:
Full test coverage for layouts

Removed some more legacy code, moved functions to different modules

Commit:
Full test coverage for placement

One line is missed, but it's a safetynet line.

Commit:
Cross construction now works for all layouts

If a horizontal or vertical qubit is not present in a placement chain,
one is randomly selected.

stefanhannie committed:
The Layout class behaves like a dictionary

Layout class inherits from abc.MutableMapping. Additional testing added.

Commit:
Full test coverage for construction

The extend method is moved to hinting, it makes much more sense there.

Commit:
Full test coverage for hinting

Random remove and extend chains are now generic options for hinting.

Commit:
Full test coverage for find_embedding

Removed some more legacy code.

Commit:
Refactor intersection, cleaner

Looks more like binning--properly uses a gridgraph as a data structure

Commit:
Starting a notebook to document various algorithms

Pretty pictures and pretty LaTeX.

stefanhannie committed:
Split construction into 3 parts, removed hinting

Construction was vague. It has been split into 3 more well defined parts:
connection, expansion, and contraction. This makes for a cleaner API.
